src/rust.rs

```diff
@@ -1257,3 +1258,55 @@ macro_rules! new_jsjitinfo_bitfield_1 {
             (($slotIndex as u32) << 22u32)
     }
 }
+
+pub struct CapturedJSStack<'a> { // lifetime inspired by typedarrays
+    cx: *mut JSContext,
+    stack: RootedGuard<'a, *mut JSObject>,
+}
+
+impl<'a> CapturedJSStack<'a> {
+    pub fn new(cx: *mut JSContext, max_frame_count: u32) -> Option<Self> { // or maybe another name, for example "capture"?
+        unsafe { // how to handle unsafe in a nice way?
+            rooted!(in(cx) let mut obj = JS_NewPlainObject(cx)); // how to overcome this problem with lifetimes?
+            let mut obj_handle = obj.handle_mut(); // if null, return None // how to handle such errors in a nice way?
+            CaptureCurrentStack(cx, obj_handle, max_frame_count); // if failed, return None
+
+            Some(CapturedJSStack {
+                cx: cx,
+                stack: obj,
+            })
+        }
+    }
+
+    pub fn as_string(&self, indent: usize) -> Option<String> {
+        // TODO where errors can occur? some example of non trivial error handling?
+        // TODO how to write tests for this code? Including the problem of obtaining some stack trace
+        // do we want to do the actual logic of conversion in this method or maybe in constructor and just return cached result?
+        unsafe {
+            let stack_handle = self.stack.handle();
+            let nullptr = 0 as *const ::std::os::raw::c_char;
```

Comment:
nit: rust has `ptr::null()` and `is_null`, so you should use them instead probably.